### PR TITLE
Flat list native onscroll

### PR DESCRIPTION
**Pull request by skv-headless**
**motivation**

use native event with FlatList

**Test plan (required)**

UIExplorer. `NativeAnimationsExample.`, `FlatListExample` (Just for demo purpose later can remove it)

### CODE REVIEW

File: Examples/UIExplorer/js/NativeAnimationsExample.js [hunk 1]
```diff
@@ -213,6 +216,43 @@ class EventExample extends React.Component {
...
+
+  render() {
+    const opacity = this.state.scrollX.interpolate({
+      inputRange: [0, 200],
+      outputRange: [1, 0],
+    });
+    return (
+      <View>
+        <Animated.View style={[ styles.block, { opacity } ]} />
+        <FlatList
```
Comment (skv-headless) [inline Examples/UIExplorer/js/NativeAnimationsExample.js]: fixed

File: Examples/UIExplorer/js/NativeAnimationsExample.js [hunk 2]
```diff
@@ -213,6 +216,43 @@ class EventExample extends React.Component {
...
+          style={{ height: 100, marginTop: 16 }}
+          scrollEventThrottle={16}
+          data={_.range(10).map((i) => ({ key: i }))}
+          renderItem={(item) => <View style={styles.block} />}
+          SeparatorComponent={() => <View style={{ width: 2 }}/>}
+          onScroll={
+            Animated.event([{
+              nativeEvent: { contentOffset: { x: this.state.scrollX } }
+            }], {
+              useNativeDriver: true,
```
Comment (skv-headless) [inline Examples/UIExplorer/js/NativeAnimationsExample.js]: fixed

**Comment by skv-headless** [comment]
cc @janicduplessis @sahrens

**Comment by sahrens** [comment]
Yeah we need to get native onScroll support working, but I don't think this is the right approach. If someone plugs in an Animated.event to onScroll and forgets to hook up the listener, everything will break.

@janicduplessis: better ideas?

File: Examples/UIExplorer/js/FlatListExample.js [hunk 3]
```diff
@@ -124,6 +130,16 @@ class FlatListExample extends React.PureComponent {
           renderItem={this._renderItemComponent}
           shouldItemUpdate={this._shouldItemUpdate}
           viewabilityConfig={VIEWABILITY_CONFIG}
+          onScroll={
+            Animated.event([{
+              nativeEvent: { contentOffset: { y: this.state.scrollY } }
+            }], {
+              useNativeDriver: true,
+              listener: (e) => {
+                this._listRef.onScroll(e);
```
Comment (sahrens) [inline Examples/UIExplorer/js/FlatListExample.js]: This is too fragile I think, unless we can come up with a robust way to detect if it's left out?

File: Libraries/CustomComponents/Lists/FlatList.js [hunk 4]
```diff
@@ -361,6 +361,10 @@ class FlatList<ItemT> extends React.PureComponent<DefaultProps, Props<ItemT>, vo
     }
   };
 
+  onScroll = (e: Object) => {
+    this._listRef._onScroll(e);
```
Comment (sahrens) [inline Libraries/CustomComponents/Lists/FlatList.js]: Should never call private functions outside the class.

**Comment by janicduplessis** [comment]
The proper way to implement this is to add a `getScrollableNode` function and use Animated.createAnimatedComponent on FlatList to create an `AnimatedFlatList` that support animated events with useNativeDriver. See https://github.com/facebook/react-native/blob/master/Libraries/CustomComponents/ListView/ListView.js#L274, this function should return the node for the ScrollVie used by FlatList.

**Comment by sahrens** [comment]
Hmm, not really a fan of that. What if `VirtualizedList` rendered `Animated.ScrollView`, passed the `AnimatedEvent` `onScroll` prop through to that `ScrollView`, and then added a listener to that `AnimatedEvent` that calls the `VirtualizedList::_onScroll` handler with a new `addListener` method?

I actually already prototyped it and it works pretty well :P I'll try to get it in soon.

**Comment by janicduplessis** [comment]
@sahrens Yea that would work well too

**Comment by sahrens** [comment]
Cool, I have a diff up internally that works pretty nicely. I introduced `Animated.forkEvent(event: ?(AnimatedEvent | Function), callback: Function): AnimatedEvent | Function` and `Animated.unforkEvent(event: ?(AnimatedEvent | Function), callback: Function): void` which work on both native and non-native events.

**Comment by skv-headless** [comment]
Seems you have working solution 👍 Will be waiting it. Just ping me if you need my help.

**Pull request closed without merging**